plugins/woocommerce/client/blocks/assets/js/blocks/product-filters/inner-blocks/taxonomy-filter/edit.tsx

```diff
@@ -22,37 +26,130 @@ import type { FilterOptionItem } from '../../types';
 import { InitialDisabled } from '../../components/initial-disabled';
 import { Notice } from '../../components/notice';
 import { getTaxonomyLabel } from './utils';
+import { sortFilterOptions } from '../../utils/sort-filter-options';
 
 const Edit = ( props: EditProps ) => {
 	const { attributes: blockAttributes } = props;
 
-	const { taxonomy, isPreview, displayStyle, showCounts, sortOrder } =
-		blockAttributes;
+	const {
+		taxonomy,
+		isPreview,
+		displayStyle,
+		showCounts,
+		sortOrder,
+		hideEmpty,
+	} = blockAttributes;
 
-	// For now, use fake data - will be replaced with real taxonomy data later
 	const [ termOptions, setTermOptions ] = useState< FilterOptionItem[] >(
 		[]
 	);
 	const [ isOptionsLoading, setIsOptionsLoading ] =
 		useState< boolean >( true );
 
+	// Fetch taxonomy terms using WordPress core data
+	const { taxonomyTerms, isTermsLoading } = useSelect(
+		( select ) => {
+			if ( ! taxonomy || isPreview ) {
+				return { taxonomyTerms: [], isTermsLoading: false };
+			}
+
+			const { getEntityRecords, hasFinishedResolution } =
+				select( coreStore );
+			const selectorArgs = [
+				'taxonomy',
+				taxonomy,
+				{
+					per_page: -1,
+					hide_empty: hideEmpty,
+					orderby: 'name',
+					order: 'asc',
+				},
+			];
+
+			return {
+				taxonomyTerms: getEntityRecords( ...selectorArgs ) || [],
+				isTermsLoading: ! hasFinishedResolution(
+					'getEntityRecords',
+					selectorArgs
+				),
+			};
+		},
+		[ taxonomy, hideEmpty, isPreview ]
+	);
+
+	// Fetch taxonomy counts using the updated useCollectionData hook
+	const { data: filteredCounts, isLoading: isFilterCountsLoading } =
+		useCollectionData( {
+			queryTaxonomy: taxonomy,
+			queryState: {},
+			isEditor: true,
+		} );
+
 	useEffect( () => {
-		setTermOptions( [
-			...termOptionsPreview.sort( ( a, b ) => {
-				switch ( sortOrder ) {
-					case 'name-asc':
-						return a.value > b.value ? 1 : -1;
-					case 'name-desc':
-						return a.value < b.value ? 1 : -1;
-					case 'count-asc':
-						return a.count > b.count ? 1 : -1;
-					default:
-						return a.count < b.count ? 1 : -1;
+		if ( isPreview ) {
+			// Use preview data for preview mode
+			setTermOptions(
+				sortFilterOptions( [ ...termOptionsPreview ], sortOrder )
+			);
+			setIsOptionsLoading( false );
+			return;
+		}
+
+		if ( isTermsLoading || isFilterCountsLoading ) {
+			setIsOptionsLoading( true );
+			return;
+		}
+
+		if ( ! taxonomy || ! taxonomyTerms.length ) {
+			setTermOptions( [] );
+			setIsOptionsLoading( false );
+			return;
+		}
```

Comment:
If the previous comment was applied, here it would be:

```suggestion
		if ( ! taxonomy || ! taxonomyTerms.length || isPreview ) {
			setTermOptions( taxonomyTerms );
			setIsOptionsLoading( false );
			return;
		}
```